processor/schemaprocessor/internal/translation/translation.go

```diff
@@ -0,0 +1,341 @@
+			*NewRevision(version, def),
+		)
+	}
+	sort.Sort(t)
+
+	t.log.Debug("Finished update")
+	return errs
+}
+
+func newTranslatorFromSchema(log *zap.Logger, schemaURL string, schemaFileSchema *ast.Schema) (*translator, error) {
+	_, target, err := GetFamilyAndVersion(schemaURL)
+	if err != nil {
+		return nil, err
+	}
+	t := &translator{
+		schemaURL: schemaURL,
+		target:    target,
+		log:       log,
+		indexes:   map[Version]int{},
+	}
+
+	if err := t.loadTranslation(schemaFileSchema); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+func newTranslatorFromReader(log *zap.Logger, schemaURL string, content io.Reader) (*translator, error) {
+	schemaFileSchema, err := encoder.Parse(content)
+	if err != nil {
+		return nil, err
+	}
+	var t *translator
+	if t, err = newTranslatorFromSchema(log, schemaURL, schemaFileSchema); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+func (t *translator) Len() int {
+	return len(t.revisions)
+}
+
+func (t *translator) Less(i, j int) bool {
+	return t.revisions[i].Version().LessThan(t.revisions[j].Version())
+}
+
+func (t *translator) Swap(i, j int) {
+	a, b := t.revisions[i].Version(), t.revisions[j].Version()
+	t.indexes[*a], t.indexes[*b] = j, i
+	t.revisions[i], t.revisions[j] = t.revisions[j], t.revisions[i]
+}
+
+func (t *translator) SupportedVersion(v *Version) bool {
+	_, ok := t.indexes[*v]
+	return ok
+}
+
+func (t *translator) ApplyAllResourceChanges(ctx context.Context, resource alias.Resource, inSchemaURL string) error {
+	_, ver, err := GetFamilyAndVersion(inSchemaURL)
+	if err != nil {
+		return err
+	}
+	it, status := t.iterator(ctx, ver)
+	for rev, more := it(); more; rev, more = it() {
+		switch status {
+		case Update:
+			err = rev.all.Apply(resource.Resource())
+			if err != nil {
+				return err
+			}
+			err = rev.resources.Apply(resource.Resource())
+			if err != nil {
+				return err
+			}
+		case Revert:
+			err = rev.resources.Rollback(resource.Resource())
+			if err != nil {
+				return err
+			}
+			err = rev.all.Rollback(resource.Resource())
+			if err != nil {
+				return err
+			}
+		}
+	}
+	resource.SetSchemaUrl(t.schemaURL)
+	return nil
+}
+
+func (t *translator) ApplyScopeLogChanges(ctx context.Context, in plog.ScopeLogs, inSchemaURL string) error {
+	_, ver, err := GetFamilyAndVersion(inSchemaURL)
+	if err != nil {
+		return err
+	}
+	it, status := t.iterator(ctx, ver)
+	if status == NoChange {
+		return nil
+	}
+	for rev, more := it(); more; rev, more = it() {
+		for l := 0; l < in.LogRecords().Len(); l++ {
+			log := in.LogRecords().At(l)
+			switch status {
+			case Update:
+				err = rev.all.Apply(log)
+				if err != nil {
+					return err
+				}
+				err = rev.logs.Apply(log)
+				if err != nil {
+					return err
+				}
+			case Revert:
+				err = rev.all.Rollback(log)
+				if err != nil {
+					return err
+				}
+				err = rev.logs.Rollback(log)
+				if err != nil {
+					return err
+				}
```

Comment:
I see in `ApplyAllResourceChanges` the `Revert` case calls `Rollback()` funcs in the reverse order compared to the order the `Update` case calls `Apply` funcs. I can't tell immediately if it matters or no, but that order reversal intuitively sounds right to me.

Why are we not doing the same here (and elsewhere)? Is it important or no? Please make it consistent (and correct - if order matters). :-)

Reply:
This was an accident.  I've double checked it and I think I fixed it all.